On admin screen, detect misconfigured WILDCARD_HOST & show warning

paulproteus opened the pull request:
This is some fairly minimal changes for showing information to the admin about a misconfigured `WILDCARD_HOST`.

Screenshot of what admin looks like when this warning is visible:

![wildcard-host-warning](https://cloud.githubusercontent.com/assets/25457/12904521/da2ac9ba-ce84-11e5-9baa-d354b58375ec.png)


Review request from:

- @neynah and/or @zarvox - please sanity-check my proposed text

Review & merge requested from:

- @kentonv

Note for Kenton: This adds a new subdomain pattern, `selftest*`.

Note that this _only_ affects `/admin/*` routes. That means that people can still get a forever-spinning grain spinner.

From my perspective, @jadeqwang and I want to limit how much time I spend on this particular bit of self-testing, so I'd like to declare making the `grain-frame` provide warning text as out-of-scope for this pull request. IMHO the most important thing is that people see this when they are doing initial setup.

Testing done:

- Set my `BASE_URL` and `WILDCARD_HOST` to point at `mouse-potato.com` which is a domain that resolves to 127.0.0.1 Verified that I saw the message.

- Fixed my `WILDCARD_HOST` and loaded the page up via `local.sandstorm.io`.

Rationale for new subdomain type:

- When I was manually testing this in Firefox, Firefox would cache the CORS result from a broken domain - and the CORS result from a broken domain is "no, CORS is not OK". Once I realized that, I realized that I needed each self-test to be on its own domain. So I created a new subdomain type.

paulproteus commented:
p.s. `mouse-potato.com` I love you so.

paulproteus commented:
Note for reviewers:

the link goes to https://docs.sandstorm.io/en/latest/administering/faq/#why-do-i-see-an-error-when-i-try-to-launch-an-app-even-when-the-sandstorm-interface-works-fine which is a directly link to the particular relevant FAQ entry

zarvox commented on shell/server/pre-meteor.js:
```diff
@@ -210,7 +210,24 @@ function checkMagic(buf, magic) {
+  inMeteor(() => {
+    if (req.method === 'GET' &&
+        req.url === '/') {
+      return serveStaticAsset(req, res, {
```
I dislike this overriding on `serveStaticAsset`.  They are not conceptually related, and I think this is the wrong level of abstraction to reuse code at.  I'd rather you just

```
res.writeHead(200, {"Content-Type": "text/plain"});
res.end("Self-test OK.");
```

for `GET`, extract a method for handling the 'OPTIONS' request (if the proper handling matches the one in `serveStaticAsset` exactly), and explicitly 405 anything else.

paulproteus replied:
I can live with that. FWIW I overloaded `staticServeAsset` because I don't
love the idea of writing copy-pasta for OPTIONS. I could extract _that_
into a function.

zarvox replied:
Note that you don't actually care about getting the same etag behavior as the static asset domain, since these are single-use host/filename pairs.  You only have one content type, you don't want to accept POSTs, and since you don't allow arbitrary user content and serve up text-plain, you don't need to worry about the rest of the CSP headers.

Maybe you need the Access-Control-Allow-Origin header, and the Content-Length header too.

paulproteus replied:
Agreed, I suppose, although personally my heart grows warmer with everything I block via CSP.

kentonv replied:
I agree this should not use the `static` subdomain, since all kinds of special-purpose behavior might apply to it in the future. I'd prefer using an independent subdomain like `self-test`. You don't need to handle `OPTIONS` -- for a `GET` request, setting `Access-Control-Allow-Origin: *` in the response is all you need.

paulproteus replied:
+1 to all this

paulproteus replied:
FWIW I went with `selftest` not `self-test` because `matchWildcardHost` doesn't permit hyphens, and I can live without a hyphen.

kentonv replied:
Oh I see you were using a separate subdomain all along, so my comment was silly. But reusing the code to handle OPTIONS still seems unnecessary since GETs don't require preflights...

neynah commented:
Wow that's scary!

Regarding copy:
- do we say Sandstorm install or Sandstorm Server? (difference is unclear to me)
- perhaps: "Once you fix the issue, reload this page." My assumption here is that there's no actual "test" other than refreshing to see if the issue is resolved?
- we could remove the documentation sentence and instead say "See documentation." or "Learn more." I think we can justify shortening it because it's clear the documentation or learn more is regarding WILDCARD_HOSTS due to the context.
- I'm unsure about "we want to help you succeed" as we don't say this anywhere else (as far as I know). And if we don't, then it makes this warning dialogue seem oddly personal. Maybe instead: "If you're still having problems, e-mail us at support@sandstorm.io." This also suggests that they should at least attempt to fix the problem themselves first.

paulproteus commented:
> do we say Sandstorm install or Sandstorm Server? (difference is unclear to me)

There's no real difference. I don't think we're consistent.

> perhaps: "Once you fix the issue, reload this page." My assumption here is that there's no actual "test" other than refreshing to see if the issue is resolved?

Yeah, the page does a diagnostic when it loads.

> we could remove the documentation sentence and instead say "See documentation." or "Learn more." I think we can justify shortening it because it's clear the documentation or learn more is regarding WILDCARD_HOSTS due to the context.

Interesting; will do that.

> I'm unsure about "we want to help you succeed" as we don't say this anywhere else (as far as I know). And if we don't, then it makes this warning dialogue seem oddly personal. Maybe instead: "If you're still having problems, e-mail us at support@sandstorm.io." This also suggests that they should at least attempt to fix the problem themselves first.

+1 I just want to make sure they're not totally put off from using Sandstorm after seeing this

neynah commented:
Oh! Another thing worth mentioning, perhaps we should state what is impacted by this issue? I know some people (myself included) ignore warnings especially if we can continue doing the thing we're doing.

dwrensha commented on shell/shared/admin.js:
```diff
@@ -46,6 +46,29 @@ var adminRoute = RouteController.extend({
 
   action: function () {
+    // Test the WILDCARD_HOST for sanity.
+    Deps.nonreactive(function() {
```
s/Deps/Tracker/ (see https://github.com/meteor/meteor/blob/devel/packages/deps/package.js)

paulproteus replied:
+1, on it, many thanks

kentonv commented on shell/shared/admin.js:
```diff
@@ -46,6 +46,29 @@ var adminRoute = RouteController.extend({
+      if (Session.get("alreadyTestedWildcardHost")) {
+        return;
+      }
+      let req = new XMLHttpRequest();
```
Meteor's `HTTP` package has a much nicer interface. Consider using it rather than using XHR directly?

paulproteus replied:
Sensible, will try that

paulproteus replied:
Using that now.

kentonv commented:
* Note that the FAQ link has an auto-generated ID which would change with any slight change to the question heading. Should we perhaps trampoline through a stable URL that will redirect to the appropriate place, so that if needed we can update the link server-side rather than have to push a new Sandstorm release?

* I wonder if changing WILDCARD_HOST and restarting the server will automatically cause clients to refresh. I think it will because Meteor.settings will have changed. So maybe we don't even need to instruct people to refresh manually?

paulproteus commented:
> Note that the FAQ link has an auto-generated ID which would change with any slight change to the question heading. Should we perhaps trampoline through a stable URL that will redirect to the appropriate place, so that if needed we can update the link server-side rather than have to push a new Sandstorm release?

I link to that from enough places that I know I'm going to have this problem in the future.

For now, my plan is (in the long run) to have docs releases correspond to Sandstorm releases, and then this link can link to the heading in the version of Sandstorm the user is running. So I'm -0 on doing  anything smart about this right now.

> I wonder if changing WILDCARD_HOST and restarting the server will automatically cause clients to refresh. I think it will because Meteor.settings will have changed. So maybe we don't even need to instruct people to refresh manually?

Possible. I can test it.

paulproteus commented:
New warning text:

![wildcard-host-warning-retr](https://cloud.githubusercontent.com/assets/25457/12905875/cc5bca36-ce8f-11e5-908a-9423870c47e0.png)

paulproteus commented:
New text, in context:

![wildcard-host-warning-retry-2](https://cloud.githubusercontent.com/assets/25457/12905932/47182300-ce90-11e5-88e0-627265c807d8.png)

paulproteus commented:
![wildcard-host-warning-retry-3](https://cloud.githubusercontent.com/assets/25457/12906022/2bbf2ce2-ce91-11e5-83c3-9d5d5d815a3a.png)

Adjusted text to say "Until you fix it", which makes it more ominous and also emphasizes that action is required.

paulproteus commented:
OK @kentonv I believe this addresses all feedback.

Note: @zarvox I chose to do 400 for all requests that aren't either `GET` to `/`, or `OPTIONS`. I could have conditional logic to `400` or `404` on `GET` of paths that aren't `/`, and additionally `405` (bad method) on non-`GET` methods, but in the interest of implementation simplicity I would prefer not to. If you would prefer I go all-in on 400 vs 404 vs 405, then I can do that.

paulproteus commented:
@dwrensha thanks again for the note about `Tracker` vs `Deps`. It's a real pleasure to work with people who provide quick feedback to help my code be more reasonable.

Same to @kentonv and @zarvox for code feedback, and @neynah for string feedback.

paulproteus commented:
Oh, also, I narrowed `Access-Control-Allow-Origin` to just the shell's origin.

paulproteus commented:
@kentonv this is ready for your re-review & merge, if you're willing!

kentonv commented on shell/server/pre-meteor.js:
```diff
@@ -210,6 +210,25 @@ function checkMagic(buf, magic) {
 }
 
+function serveSelfTest(req, res) {
+  inMeteor(() => {
```
This inMeteor() is unnecessary since you aren't using any meteor APIs in the code within.

paulproteus replied:
But it does mean that exceptions won't bubble-up to the top and destroy the entire process, which otherwise they would do, I believe.

If I'm mistaken, let me know, but I'm gently in favor of keeping it given the above.

paulproteus replied:
Upon re-reading this comment, I see that `inMeteor()` is something that we wrote, not something that comes with Meteor. Since `try/except` is all I need, I'll switch to that.

kentonv commented on shell/server/pre-meteor.js:
```diff
@@ -210,6 +210,25 @@ function checkMagic(buf, magic) {
+      res.writeHead(200, {
+        'Content-Type': 'text/plain',
+        'Content-Length': content.length,
+        'Access-Control-Allow-Origin': process.env.ROOT_URL,
```
You could make this `*`, which would avoid the need to change it later when we support multiple `ROOT_URL`s. I don't see any danger in allowing all origins.

paulproteus replied:
In theory allowing all origins is an information leak that this entity is running a Sandstorm at their domain. If we don't need `*` yet, then I'm happier to limit this to just what is needed. I can allow `*` if you very much prefer, though.

kentonv commented on shell/server/pre-meteor.js:
```diff
@@ -471,6 +490,12 @@ Meteor.startup(() => {
         return;
       }
 
+      if (id.match(/^selftest/)) {
```
Technically, a valid grain host could begin with "selftest". Note that "static" is not a valid grain host because grain host IDs are always longer than that, but for "selftest" you're only matching a prefix. I think it would make sense to permit hyphens in wildcard IDs so that you can make this "selftest-*", which is never a valid grain host.

kentonv replied:
(We may want to use this sort of pattern for other things in the future, e.g. API endpoints being `api-*`.)

kentonv replied:
FWIW I think making hyphens be allowed is a one-char change to a regex in db.js.

paulproteus replied:
I always use `Random.hexString(20)` in the "client" code for this, so I'm +1 on me changing this to be a different length, too.

I'd prefer the length-based way to differentiate the selftest hosts, but I make it use a hyphen if you prefer that.

paulproteus replied:
OK, I'll support hyphens! Done in a follow-up commit.

paulproteus commented:
@kentonv in your last review, you had these requests.

- Support `Access-Control-Allow-Origin: *`. However, I think it's OK to only allow the shell's endpoint. I opened https://github.com/sandstorm-io/sandstorm/pull/1490 if you prefer to merge it that way.

- Stop using `inMeteor()`: Done. The only thing I needed was try/catch.

- Use a hyphen to clarify `selftest-*` wildcards from other wildcards. Done.

Ready for your fresh review/merge. I personally prefer this branch over #1490 but either way works for me.

kentonv commented:
Hmm, apparently we need to update the kernel headers on our build machine.

kentonv commented:
Garply, retest this please.
